pre-aio-node-with-zuul-ref.sh

```diff
@@ -0,0 +1,157 @@
+#!/bin/bash
+
+set -ex
+refspec="$1"
+ZUUL_CHANGES="$2"
+keep_env="$3"
+
+FETCH_URL=http://10.71.212.50/p/
+ROOT_DIR=/opt/stack/
+DEVSTACK_TOOL=/opt/stack/devstack
+NOVA_DIR=/opt/stack/nova
+TEMPEST_DIR=/opt/stack/tempest
+TMP_MERGE_BRANCH=tmp_branch_for_vgpu_patches
+DEST=${DEST:-"/opt/stack/openstack"}
+NOVA_CONF_PATH=/etc/nova
+SAMPLE_IMAGE=cirros-0.3.5-x86_64-disk
+VM_NAME="testVM"
+# Set up internal variables
+_SSH_OPTIONS="\
+    -o BatchMode=yes \
+    -o StrictHostKeyChecking=no \
+    -o UserKnownHostsFile=/dev/null \
+    -i /opt/stack/.ssh/id_rsa"
+
+server_info=$(grep "VNCSERVER_PROXYCLIENT_ADDRESS" local.conf)
+server_info=${server_info#*\"}
+server_info=${server_info%\"}
+echo "servier ip:"
+echo $server_info
+
+function on_xenserver() {
+    ssh $_SSH_OPTIONS "root@$server_info" bash -s --
+}
+
+cd $DEVSTACK_TOOL/
+./clean.sh
+git checkout master
+git pull --ff origin master
+
+sed -i '/[[local|localrc]]/a DEST=/opt/stack/openstack' ${DEVSTACK_TOOL}/local.conf
+
+IFS='^' read -r -a ZUUL_PROJ_CHANGE <<< "$ZUUL_CHANGES"
+related_proj=()
+for proj in ${ZUUL_PROJ_CHANGE[@]}
+do
+    proj=${proj%%:*}
+    proj=${proj##*/}
+    related_proj+=($proj)
+done
+unique_proj_list=$(echo "${related_proj[@]}" | tr ' ' '\n' | nl | sort -u -k2 | sort -n | cut -f2-)
```

Comment:
line 42 - 50 can be simplified as:
PROJECT_LIST=$(echo $ZUUL_CHANGES | tr '^' '\n' | cut -d: -f1 | sort | uniq)

Reply:
Awesome command~